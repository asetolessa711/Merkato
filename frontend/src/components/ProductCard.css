.product-card {
  width: 180px;
  border-radius: 14px;
  box-shadow: 0 2px 12px rgba(0,0,0,0.07);
  transition: transform 0.2s, box-shadow 0.2s;
  background: #fff;
  cursor: pointer;
  overflow: hidden;
  position: relative;
}

.product-card:hover, .product-card.active {
  transform: scale(1.07);
  box-shadow: 0 6px 24px rgba(0,0,0,0.13);
  z-index: 2;
}

.product-image-wrapper {
  position: relative;
  width: 100%;
  padding-top: 100%; /* 1:1 aspect ratio */
  background: #f8f8f8;
}

.product-image-wrapper img {
  position: absolute;
  top: 10%;
  left: 10%;
  width: 80%;
  height: 80%;
  object-fit: contain;
  border-radius: 10px;
}

.favorite-btn {
  position: absolute;
  top: 8px;
  right: 8px;
  background: rgba(255,255,255,0.8);
  border: none;
  border-radius: 50%;
  font-size: 1.1rem;
  cursor: pointer;
  color: #e17055;
  transition: background 0.2s;
}

.favorite-btn:hover {
  background: #ffeaa7;
}

.discount-badge {
  position: absolute;
  top: 8px;
  left: 8px;
  background: #e74c3c;
  color: #fff;
  border-radius: 6px;
  padding: 2px 8px;
  font-size: 0.8rem;
}

.product-info {
  padding: 10px 8px 14px 8px;
  text-align: center;
}

.product-info h4 {
  font-size: 1rem;
  font-weight: 600;
  margin: 0 0 4px 0;
}

.product-meta {
  display: flex;
  justify-content: center;
  gap: 8px;
  align-items: center;
  margin-bottom: 8px;
}

.price {
  color: #00b894;
  font-weight: bold;
  font-size: 1.1rem;
}

.stars {
  color: #f1c40f;
  font-size: 0.95rem;
}

.shop-btn {
  background: #0984e3;
  color: #fff;
  border: none;
  border-radius: 8px;
  padding: 6px 14px;
  font-weight: bold;
  cursor: pointer;
  transition: background 0.2s;
}

.shop-btn:hover {
  background: #00b894;
}

/* Sizes */
.size-sm { width: 160px; font-size: 0.85rem; }
.size-md { width: 220px; font-size: 1rem; }
.size-lg { width: 280px; font-size: 1.1rem; }

/* Color swatches */
.color-preview {
  display: flex;
  gap: 6px;
  margin-top: 6px;
}

.color-swatch {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  border: 1px solid #ccc;
}

.color-swatch:hover {
  transform: scale(1.1);
  box-shadow: 0 0 4px rgba(0,0,0,0.2);
}

/* Universal Grid Layout for Product Cards */
.products-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(220px, 1fr));
  gap: 24px;
  margin-top: 24px;
  padding: 10px 16px;
  justify-content: center;
  align-items: stretch;
}

@media (max-width: 600px) {
  .products-grid {
    grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
    gap: 16px;
    padding: 8px;
  }
}
