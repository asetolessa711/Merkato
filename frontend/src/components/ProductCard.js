import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './ProductCard.css';

function ProductCard({
  product,
  type = 'standard',
  size = 'md', // sm | md | lg | xl
  colorOptions = [], // e.g. ['#FF0000', '#000']
  onAddToCart,
}) {
  const isDeal = type === 'deal' || product.promotion?.isPromoted || product.discount > 0;
  const discountText = product.discount > 0 ? `-${product.discount}%` : '';
  const finalPrice = product.price.toFixed(2);
  const imageSrc = product.image || '/images/default-product.png';
  const theme = product.theme || 'mint'; // fallback
  const isOutOfStock = product.stock === 0;

  return (
    <div className={`product-card theme-${theme} size-${size} ${isDeal ? 'deal-card' : ''} ${isOutOfStock ? 'out-of-stock-card' : ''}`}>
      {/* Product Image */}
      <div className="product-card-image-container">
        <img src={imageSrc} alt={product.name} loading="lazy" />
      </div>
      {/* Product Info */}
      <div className="product-info">
        <h3 className="product-title">{product.name}</h3>
        <div className="product-price">
          <span>${finalPrice}</span>
        </div>
        {/* Product Description */}
        {product.description && (
          <div className="product-description">{product.description}</div>
        )}
        {/* Add to Cart Button */}
        <button
          type="button"
          onClick={() => !isOutOfStock && typeof onAddToCart === 'function' ? onAddToCart(product) : undefined}
          disabled={isOutOfStock}
          aria-label="Add to Cart"
        >
          Add to Cart
        </button>
        {/* CTA */}
        <Link to={`/product/${product._id}`} className="btn-small">
          Shop Now
        </Link>
      </div>
    </div>
  );
}

ProductCard.propTypes = {
  product: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    image: PropTypes.string,
    price: PropTypes.number.isRequired,
    discount: PropTypes.number,
    theme: PropTypes.string,
    promotion: PropTypes.shape({
      isPromoted: PropTypes.bool,
      badgeText: PropTypes.string,
    }),
    vendor: PropTypes.shape({
      name: PropTypes.string,
    }),
    stock: PropTypes.number,
  }).isRequired,
  type: PropTypes.oneOf(['standard', 'deal']),
  size: PropTypes.oneOf(['sm', 'md', 'lg', 'xl']),
  colorOptions: PropTypes.arrayOf(PropTypes.string),
  onAddToCart: PropTypes.func,
};

export default ProductCard;
