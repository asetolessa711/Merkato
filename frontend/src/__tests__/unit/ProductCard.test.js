import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ProductCard from '../../components/ProductCard';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';

// Mock product data
const product = {
  _id: '123',
  name: 'Test Product',
  price: 49.99,
  description: 'A nice product for testing',
  image: 'test-product.jpg',
};

describe('\ud83d\udecd\ufe0f ProductCard Component', () => {
  test('renders product name and price', () => {
    render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );
    expect(screen.getByText(/test product/i)).toBeInTheDocument();
    expect(screen.getByText(/\$?49\.99/i)).toBeInTheDocument();
  });

  test('displays product image with alt text', () => {
    render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );
    const img = screen.getByRole('img');
    expect(img).toHaveAttribute('src', expect.stringContaining('test-product.jpg'));
    expect(img).toHaveAttribute('alt', expect.stringMatching(/test product/i));
  });

  test('calls add-to-cart when button clicked', () => {
    const onAddToCart = jest.fn();
    render(
      <MemoryRouter>
        <ProductCard product={product} onAddToCart={onAddToCart} />
      </MemoryRouter>
    );
    const btn = screen.getByRole('button', { name: /add to cart/i });
    fireEvent.click(btn);
    expect(onAddToCart).toHaveBeenCalledTimes(1);
    expect(onAddToCart).toHaveBeenCalledWith(product);
  });

  test('disables add-to-cart button if out of stock', () => {
    render(
      <MemoryRouter>
        <ProductCard product={{ ...product, stock: 0 }} />
      </MemoryRouter>
    );
    const btn = screen.getByRole('button', { name: /add to cart/i });
    expect(btn).toBeDisabled();
  });

  test('shows product description', () => {
    render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );
    expect(screen.getByText(/a nice product for testing/i)).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
