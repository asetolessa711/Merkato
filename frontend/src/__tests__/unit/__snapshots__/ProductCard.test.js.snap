// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`üõçÔ∏è ProductCard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product-card theme-mint size-md  "
    data-testid="product-card"
  >
    <div
      class="product-card-image-container"
    >
      <img
        alt="Test Product"
        loading="lazy"
        src="test-product.jpg"
      />
    </div>
    <div
      class="product-info"
    >
      <h3
        class="product-title"
      >
        <a
          href="/product/123"
        >
          Test Product
        </a>
      </h3>
      <div
        class="product-price"
      >
        <span>
          $49.99
        </span>
      </div>
      <div
        class="product-description"
      >
        A nice product for testing
      </div>
      <button
        aria-label="Add to Cart"
        data-testid="add-to-cart-btn"
        type="button"
      >
        Add to Cart
      </button>
      <a
        class="btn-small"
        href="/product/123"
      >
        Shop Now
      </a>
    </div>
  </div>
</DocumentFragment>
`;
